### PR TITLE
Refactoring : part I

### PR DESCRIPTION
**Pull request by frapac**
This is a first draft for the refactoring of ExaPF. Following our discussion, this first PR attempts at:
- move all the structure of the problem in the submodule `PowerSystem` and creates a proper directory for this module. 
- introduce an abstract type `AbstractPowerSystem` to allow different formulation. Each object inheriting from `AbstractPowerSystem` must define the mappings `get_{x,u,p}` and `retrieve_physics`. Basically, `AbstractPowerSystem` is now the de facto interface between the physical and the mathematical layers
- once the mathematical formulation specified, we could implement the different callbacks (cost, constraints). As these callbacks depend on the formulation used (we currently have only polar formulation, but might want to implement other formulations), this PR moves all formulation dependent functions in a new file `formulations.jl`. 

Any feedback is welcome (especially about the naming, not sure it is explicit enough)!

### CODE REVIEW

**Comment by frapac** [comment]
The more I think about it, the more I think we should see `PowerSystem` as an independent module, which serves only the purpose of specifying the physical configuration of the system. Then, we could specify elsewhere how we want to use this information to formulate a mathematical problem.

**Comment by dmaldona** [comment]
@frapac excellent PR proposal. I completely agree with refactoring and moving equations and such out of ExaPF.jl

File: src/PowerSystem/PowerSystem.jl [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+using SparseArrays
+using ..ExaPF: ParsePSSE, ParseMAT, IndexSet, Spmat
+
+import Base: show
+
+const PQ_BUS_TYPE = 1
+const PV_BUS_TYPE = 2
+const REF_BUS_TYPE  = 3
+
+abstract type AbstractNetworkAttribute end
```
Comment (frapac) [inline src/PowerSystem/PowerSystem.jl:13]: At some point, we should discuss the attributes we need here, for the power system.

File: src/PowerSystem/PowerSystem.jl [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+abstract type AbstractIndexing end
+struct PVIndexes <: AbstractIndexing end
+struct PQIndexes <: AbstractIndexing end
+struct SlackIndexes <: AbstractIndexing end
+struct GeneratorIndexes <: AbstractIndexing end
+struct PVGeneratorIndexes <: AbstractIndexing end
+struct SlackGeneratorIndexes <: AbstractIndexing end
+
+# TODO: replace const *_BUS_TYPE with this enum
+@enum BusType begin
```
Comment (frapac) [inline src/PowerSystem/PowerSystem.jl:38]: Still need to figure out whether we should use a `enum` or proper `struct` with careful subtyping.

File: src/PowerSystem/power_network.jl [hunk 3]
```diff
@@ -78,6 +66,25 @@ struct PowerNetwork
...
 end
 
+get(pf::PowerNetwork, ::NumberOfBuses) = pf.nbus
+get(pf::PowerNetwork, ::NumberOfLines) = size(pf.data["branch"], 1)
+get(pf::PowerNetwork, ::NumberOfGenerators) = pf.ngen
+get(pf::PowerNetwork, ::NumberOfPVBuses) = length(pf.pv)
+get(pf::PowerNetwork, ::NumberOfPQBuses) = length(pf.pq)
+get(pf::PowerNetwork, ::NumberOfSlackBuses) = length(pf.ref)
+
+function get(pf::PowerNetwork, ::GeneratorIndexes)
```
Comment (frapac) [inline src/PowerSystem/power_network.jl:76]: At some point, we may want to leverage this function to support multiple generators by bus.

File: src/ad.jl [hunk 4]
```diff
@@ -25,7 +25,7 @@ struct StateJacobianAD <: AbstractJacobianAD
     varx
     t1svarx
     map
-    function StateJacobianAD(residualFunction, F, v_m, v_a, ybus_re, ybus_im, pinj, qinj, pv, pq, ref, nbus)
+    function StateJacobianAD(F, v_m, v_a, ybus_re, ybus_im, pinj, qinj, pv, pq, ref, nbus)
```
Comment (frapac) [inline src/ad.jl:28]: I remove `residualFunction`, as we do not need it. Also, we do need `F` as well

File: src/evaluators.jl [hunk 5]
```diff
@@ -0,0 +1,122 @@
+
```
Comment (frapac) [inline src/evaluators.jl:1]: This file implements the `ADFactory` we discussed last Friday

File: src/evaluators.jl [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+# WIP: definition of AD factory
+abstract type AbstractADFactory end
+
+struct ADFactory <: AbstractADFactory
+    Jgₓ::AD.StateJacobianAD
+    Jgᵤ::AD.DesignJacobianAD
+end
+
+abstract type AbstractNLPEvaluator end
```
Comment (frapac) [inline src/evaluators.jl:12]: At some point, we may want a `FullSpaceEvaluator` as well

File: src/evaluators.jl [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    u_min, u_max = bounds(model, Control())
+    x_min, x_max = bounds(model, State())
+    return ReducedSpaceEvaluator(model, x, p, x_min, x_max, u_min, u_max,
+                                 ad, ε_tol)
+end
+
+n_variables(nlp::ReducedSpaceEvaluator) = length(nlp.u_min)
+# n_constraints(nlp::ReducedSpaceEvaluator) = length(nlp.u)
+
+function update!(nlp::ReducedSpaceEvaluator, u)
```
Comment (frapac) [inline src/evaluators.jl:40]: Each time we have a new `u`, we need to recompute the corresponding `x`

File: src/evaluators.jl [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+    ∇fₓ = fdCdx(xₖ)
+    ∇fᵤ = fdCdu(u)
+    # Update adjoint
+    λₖ = -(∇gₓ'\∇fₓ)
+    # compute reduced gradient
+    g .= ∇fᵤ + (∇gᵤ')*λₖ
+    return nothing
+end
+
+function constraint!(nlp::ReducedSpaceEvaluator, cons, u)
```
Comment (frapac) [inline src/evaluators.jl:73]: The remaining of this file is not functioning yet

File: src/formulations.jl [hunk 9]
```diff
@@ -0,0 +1,286 @@
+
```
Comment (frapac) [inline src/formulations.jl:1]: This file is a buffer, where I move the implementation of the functions we have previously in `src/ExaPF.jl`. We should discuss what we need precisely to get rid of the dead code

File: src/models/models.jl [hunk 10]
```diff
@@ -0,0 +1,185 @@
+export PolarForm, get, bounds, powerflow
```
Comment (frapac) [inline src/models/models.jl:1]: Implement all the abstractions needed to implement new formulations (polar, rectangular) in a modular way
We should discuss here what to export exactly

File: src/models/models.jl [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+export State, Control, Parameters, NumberOfState, NumberOfControl
+
```
Comment (frapac) [inline src/models/models.jl:3]: Also: I welcome any feedback on the naming. I am not sure I am explicit enough there :-)

File: src/models/polar.jl [hunk 12]
```diff
@@ -0,0 +1,509 @@
+# Polar formulation
```
Comment (frapac) [inline src/models/polar.jl:1]: Implement the polar formulation, with the new abstraction

File: src/models/polar.jl [hunk 13]
```diff
@@ -0,0 +1,509 @@
...
+    x_max::VT
+    u_min::VT
+    u_max::VT
+    # costs
+    costs_coefficients::AT
+    # Constants
+    active_load::VT
+    reactive_load::VT
+    # struct
+    ybus_re::Spmat
```
Comment (frapac) [inline src/models/polar.jl:16]: We now store `ybus_{re,im}` once and for all

File: src/models/polar.jl [hunk 14]
```diff
@@ -0,0 +1,509 @@
...
+    ybus_im::Spmat
+    AT::Type
```
Comment (frapac) [inline src/models/polar.jl:18]: I do not think we need that

File: src/models/polar.jl [hunk 15]
```diff
@@ -0,0 +1,509 @@
...
+    network::PS.PowerNetwork
+    device::Device
+    x_min::VT
+    x_max::VT
+    u_min::VT
+    u_max::VT
+    # costs
+    costs_coefficients::AT
+    # Constants
+    active_load::VT
```
Comment (frapac) [inline src/models/polar.jl:13]: Move active and reactive loads in the format we want (`CuVector` when we are working on the GPU)

File: src/models/polar.jl [hunk 16]
```diff
@@ -0,0 +1,509 @@
...
+        M = SparseMatrixCSC
+        AT = Array
+    elseif isa(device, CUDADevice)
+        VT = CuVector
+        M = CuSparseMatrixCSR
+        AT = CuArray
+    end
+    ybus_re, ybus_im = Spmat{VT}(pf.Ybus)
+    coefs = PS.get_costs_coefficients(pf) |> AT
+    u_min, u_max, x_min, x_max, p_min, p_max = PS.get_bound_constraints(pf)
```
Comment (frapac) [inline src/models/polar.jl:33]: Still need to work on the abstraction there. `PowerSystem` should not be aware of the indexing of `u` and `x` (as done here), as we want to implement different formulations based on `PowerSystem`

File: src/models/polar.jl [hunk 17]
```diff
@@ -0,0 +1,509 @@
...
+    p[nref + npq + 1:nref + 2*npq] = qbus[polar.network.pq]
+    return p
+end
+
+# Bridge with buses' attributes
+function get(polar::PolarForm{T, VT, AT}, ::PS.Buses, ::PS.VoltageMagnitude, x, u, p; V=eltype(x)) where {T, VT, AT}
+    nbus = PS.get(polar.network, PS.NumberOfBuses())
+    npv = PS.get(polar.network, PS.NumberOfPVBuses())
+    npq = PS.get(polar.network, PS.NumberOfPQBuses())
+    nref = PS.get(polar.network, PS.NumberOfSlackBuses())
```
Comment (frapac) [inline src/models/polar.jl:132]: We need to have a discussion here: the code could become a bit verbose with the current implementation, so we should decide if we want to be more concise or not.

File: src/models/polar.jl [hunk 18]
```diff
@@ -0,0 +1,509 @@
...
+    MT = polar.AT
+    vmag = MT{V, 1}(undef, nbus)
```
Comment (frapac) [inline src/models/polar.jl:134]: Yeah, this is ugly... Unfortunately that the best I found so far to support both `DualNumber` and `CuVector`. Any suggestion to improve that welcome!

File: src/models/polar.jl [hunk 19]
```diff
@@ -0,0 +1,509 @@
...
+    vmag[polar.network.pq] = x[1:npq]
+    vmag[polar.network.ref] = u[1:nref]
+    vmag[polar.network.pv] = u[nref + npv + 1:nref + 2*npv]
+    return vmag
+end
+function get(polar::PolarForm{T, VT, AT}, ::PS.Buses, ::PS.VoltageAngle, x, u, p; V=eltype(x)) where {T, VT, AT}
```
Comment (frapac) [inline src/models/polar.jl:140]: Note that we still need `V` in the signature

File: src/models/polar.jl [hunk 20]
```diff
@@ -0,0 +1,509 @@
...
+
+function bounds(polar::PolarForm{T, VT, AT}, ::State) where {T, VT, AT}
+    npq = PS.get(polar.network, PS.NumberOfPQBuses())
+    return polar.x_min[1:npq], polar.x_max[1:npq]
+end
+function bounds(polar::PolarForm{T, VT, AT}, ::Control) where {T, VT, AT}
+    return polar.u_min, polar.u_max
+end
+
+function get_network_state(polar::PolarForm{T, VT, AT}, x, u, p; V=Float64) where {T, VT, AT}
```
Comment (frapac) [inline src/models/polar.jl:215]: Equivalent to the former `retrieve_physics`

File: src/models/polar.jl [hunk 21]
```diff
@@ -0,0 +1,509 @@
...
+    fill!(F, 0.0)
+    residualFunction_polar!(F, Vm, Va,
+                                  polar.ybus_re,
+                                  polar.ybus_im,
+                                  pbus, qbus, pv, pq, nbus)
+    return F
+end
+
+# TODO: find better naming
+function init_ad_factory(polar::PolarForm{T, VT, AT}, x, u, p) where {T, VT, AT}
```
Comment (frapac) [inline src/models/polar.jl:273]: Instantiate AD once for all

File: src/models/polar.jl [hunk 22]
```diff
@@ -0,0 +1,509 @@
...
+    fill!(F, zero(T))
+    # Build the AD Jacobian structure
+    stateJacobianAD = AD.StateJacobianAD(F, Vm, Va,
+                                         polar.ybus_re, polar.ybus_im, pbus, qbus, pv, pq, ref, nbus)
+    designJacobianAD = AD.DesignJacobianAD(F, Vm, Va,
+                                           polar.ybus_re, polar.ybus_im, pbus, qbus, pv, pq, ref, nbus)
+    return stateJacobianAD, designJacobianAD
+end
+
+function jacobian(polar::PolarForm, jac::AD.AbstractJacobianAD, x, u, p)
```
Comment (frapac) [inline src/models/polar.jl:295]: Evaluate Jacobian of powerflow equations

File: src/models/polar.jl [hunk 23]
```diff
@@ -0,0 +1,509 @@
...
+    pv = convert(polar.AT{Int, 1}, polar.network.pv)
+    pq = convert(polar.AT{Int, 1}, polar.network.pq)
+    # Network state
+    Vm, Va, pbus, qbus = get_network_state(polar, x, u, p)
+    AD.residualJacobianAD!(jac, residualFunction_polar!, Vm, Va,
+                           polar.ybus_re, polar.ybus_im, pbus, qbus, pv, pq, ref, nbus, TIMER)
+    return jac.J
+end
+
+function powerflow(
```
Comment (frapac) [inline src/models/polar.jl:309]: The new `solve` function, working now directly with `PolarForm`. I think we could go further in the refactoring there

File: src/models/polar.jl [hunk 24]
```diff
@@ -0,0 +1,509 @@
...
+    ref = convert(polar.AT{Int, 1}, polar.network.ref)
+    pv = convert(polar.AT{Int, 1}, polar.network.pv)
+    pq = convert(polar.AT{Int, 1}, polar.network.pq)
+
+    # iteration variables
+    iter = 0
+    converged = false
+
+    # indices
+    j1 = 1
```
Comment (frapac) [inline src/models/polar.jl:344]: At some point, the indexing SHOULD be specified in the formulation `PolarForm` directly. But I am still figuring out a clean way to do so

File: src/models/polar.jl [hunk 25]
```diff
@@ -0,0 +1,509 @@
...
+        n_iters = Iterative.ldiv!(dx, J, F, solver, preconditioner, TIMER)
+        push!(linsol_iters, n_iters)
+        # Sometimes it is better to move backward
+        dx .= -dx
+
+        # update voltage
+        @timeit TIMER "Update voltage" begin
+            if (npv != 0)
+                # Va[pv] .= Va[pv] .+ dx[j5:j6]
+                Vapv .= Vapv .+ dx56
```
Comment (frapac) [inline src/models/polar.jl:399]: Idem. This step depends on the formulation we are using

File: src/models/polar.jl [hunk 26]
```diff
@@ -0,0 +1,509 @@
...
+    power_generations = get(polar, PS.Generator(), PS.ActivePower(), x, u, p; V=V)
+    c0 = polar.costs_coefficients[:, 2]
+    c1 = polar.costs_coefficients[:, 3]
+    c2 = polar.costs_coefficients[:, 4]
+    # Return quadratic cost
+    cost = sum(c0 .+ c1 .* power_generations + c2 .* power_generations.^2)
+    return cost
+end
+
+# Generic inequality constraints
```
Comment (frapac) [inline src/models/polar.jl:465]: Implement generic constraints afterward

File: src/PowerSystem/PowerSystem.jl [hunk 27]
```diff
@@ -0,0 +1,70 @@
...
+const REF_BUS_TYPE  = 3
+
+abstract type AbstractNetworkAttribute end
+struct NumberOfBuses <: AbstractNetworkAttribute end
+struct NumberOfLines <: AbstractNetworkAttribute end
+struct NumberOfGenerators <: AbstractNetworkAttribute end
+struct NumberOfPVBuses <: AbstractNetworkAttribute end
+struct NumberOfPQBuses <: AbstractNetworkAttribute end
+struct NumberOfSlackBuses <: AbstractNetworkAttribute end
+
```
Comment (dmaldona) [inline src/PowerSystem/PowerSystem.jl:20]: So, my understanding is that these structs are used to specify the "get" call. So, e.g., if I do

`get(pf, NumberOfBuses)`

then Julia knows that I am referring to:

`get(pf::PowerNetwork, ::NumberOfBuses) = pf.nbus`

File: src/PowerSystem/PowerSystem.jl [hunk 28]
```diff
@@ -0,0 +1,70 @@
...
+struct NetworkState{VT<:AbstractVector}
+    vmag::VT
+    vang::VT
+    pbus::VT
+    qbus::VT
+end
+
+abstract type AbstractPowerSystem end
+# Templating
+function get(::AbstractPowerSystem, ::AbstractNetworkAttribute) end
```
Comment (dmaldona) [inline src/PowerSystem/PowerSystem.jl:65]: Why is this necessary?

File: src/evaluators.jl (hunk 7 above)
Comment (dmaldona) [reply src/evaluators.jl:40]: This whole module is excellent.

File: src/models/models.jl (hunk 11 above)
Comment (dmaldona) [reply src/models/models.jl:3]: I think you are being explicit enough.

File: src/PowerSystem/PowerSystem.jl (hunk 27 above)
Comment (frapac) [reply src/PowerSystem/PowerSystem.jl:20]: that's correct!

**Review by michel2323 — APPROVED** [review]
After the walkthrough this morning I still haven't filly grasped the changes, but it looks very nice to me. I ran the tests and they all pass. So let's merge!

**Review by dmaldona — APPROVED** [review]
